v2/api/network/customizations/virtual_network_extensions.go

```diff
@@ -0,0 +1,200 @@
+		transformed, err = transformToARM(ctx, &subnet, subnetGVK, kubeClient, resolver)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to transform subnet %s/%s", subnet.GetNamespace(), subnet.GetName())
+		}
+		armSubnets = append(armSubnets, transformed)
+	}
+
+	log.V(Info).Info("Found subnets to include on VNET", "count", len(armSubnets), "names", getSubnetNames(armSubnets))
+
+	err = fuzzySetSubnets(armObj.Spec(), armSubnets)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to set subnets")
+	}
+
+	return armObj, nil
+}
+
+func getSubnetNames(subnets []genruntime.ARMResourceSpec) []string {
+	var result []string
+	for _, s := range subnets {
+		result = append(result, s.GetName())
+	}
+
+	return result
+}
+
+func getSubnetGVK(vnet genruntime.MetaObject) schema.GroupVersionKind {
+	gvk := genruntime.GetOriginalGVK(vnet)
+	gvk.Kind = reflect.TypeOf(network.VirtualNetworksSubnet{}).Name() // "VirtualNetworksSubnet"
+
+	return gvk
+}
+
+func transformToARM(
+	ctx context.Context,
+	obj genruntime.MetaObject,
+	gvk schema.GroupVersionKind,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver) (genruntime.ARMResourceSpec, error) {
+
+	spec, err := genruntime.GetVersionedSpecFromGVK(obj, kubeClient.Scheme(), gvk)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get spec from %s/%s", obj.GetNamespace(), obj.GetName())
+	}
+
+	armTransformer, ok := spec.(genruntime.ARMTransformer)
+	if !ok {
+		return nil, errors.Errorf("spec was of type %T which doesn't implement genruntime.ArmTransformer", spec)
+	}
+
+	_, resolvedDetails, err := resolver.ResolveAll(ctx, obj)
+	if err != nil {
+		return nil, arm.ClassifyResolverError(err)
+	}
+
+	armSpec, err := armTransformer.ConvertToARM(resolvedDetails)
+	if err != nil {
+		return nil, errors.Wrapf(err, "transforming resource %s to ARM", obj.GetName())
+	}
+
+	typedArmSpec, ok := armSpec.(genruntime.ARMResourceSpec)
+	if !ok {
+		return nil, errors.Errorf("casting armSpec of type %T to genruntime.ARMResourceSpec", armSpec)
+	}
+
+	return typedArmSpec, nil
+}
+
+// TODO: When we move to Swagger as the source of truth, the type for vnet.properties.subnets and subnet.properties
+// TODO: may be the same, so we can do away with the JSON serialization part of this assignment.
+// fuzzySetSubnets assigns a collection of subnets to the subnets property of the vnet. Since there are
+// many possible ARM API versions and we don't know which one we're using, we cannot do this statically.
+// To make matters even more horrible, the type used in the vnet.properties.subnets property is not the same
+// type as used for subnet.properties (although structurally they are basically the same). To overcome this
+// we JSON serialize the subnet and deserialize it into the vnet.properties.subnets field.
+func fuzzySetSubnets(vnet genruntime.ARMResourceSpec, subnets []genruntime.ARMResourceSpec) error {
+	// Here be dragons
+	vnetValue := reflect.ValueOf(vnet)
```

Comment:
Why do we need reflection here?

Why not just cast to the expected ARM type and use that?

Reply:
The problem is that somebody could create a VNET with APIVersion 2021-04-01, and a subnet with APIVersion 2021-06-01.

So the storage type for the VNET would be 06-01storage, but the ARM type (which is the type we're dealing with here) is determined at runtime as 2021-04-01. That means the type we're actually working on isn't determinable at compile time (it's a function of what they created the resource with).

Reply:
oh actually I explain this in the comment above:
> // fuzzySetSubnets assigns a collection of subnets to the subnets property of the vnet. Since there are
// many possible ARM API versions and we don't know which one we're using, we cannot do this statically.
// To make matters even more horrible, the type used in the vnet.properties.subnets property is not the same
// type as used for subnet.properties (although structurally they are basically the same). To overcome this
// we JSON serialize the subnet and deserialize it into the vnet.properties.subnets field.

Reply:
Oh, I understood the "different types in the same package" bit but missed the rest. All good.